diff --git a/setcifsacl.1 b/setcifsacl.1
index 8550201..1f66ddf 100644
--- a/setcifsacl.1
+++ b/setcifsacl.1
@@ -97,7 +97,7 @@ Set an ACL
 .br
 setcifsacl -S "ACL:CIFSTESTDOM\\Administrator:0x0/0x0/FULL,
 .br
-ACL:CIFSTESTDOM\\user2:0x0/0x0/FULL," <file_name>
+ACL:CIFSTESTDOM\\user2:0x0/0x0/FULL" <file_name>
 .PP
 .SH "AUTHOR"
 .PP
diff --git a/setcifsacl.c b/setcifsacl.c
index e1713c3..035cea4 100644
--- a/setcifsacl.c
+++ b/setcifsacl.c
@@ -694,8 +694,10 @@ get_numcaces(const char *aces)
 	const char *current;
 
 	current = aces;
-	while((current = strchr(current, ',')))
+	while((current = strchr(current, ','))) {
+		++current;
 		++num;
+	}
 
 	return num;
 }
